README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- Copyright (C) 2021 Merck Sharp & Dohme Corp. a subsidiary of Merck & Co., Inc., Kenilworth, NJ, USA.

 This file is part of bayesiansprt.

     bayesiansprt is free software: you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published by
     the Free Software Foundation, either version 3 of the License, or
     (at your option) any later version.

     bayesiansprt is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
     GNU General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with bayesiansprt.  If not, see <https://www.gnu.org/licenses/>. -->

# bayesiansprt

<!-- badges: start -->
<!-- badges: end -->

The goal of `bayesiansprt` (under GPL-3 licence) is to provide the
results for sequential probability ratio test under frequentist and
Bayesian setup.

## Installation

You can install the released version of `bayesiansprt` from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("bayesiansprt")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(bayesiansprt)
## basic example code
method = "Bayes"
n.sim = 10000
n.event = 60
n.start = 30
n.increment = 5
ve.null = 0.3
ve.alt = 0.75
type1 = 0.025
power = 0.9
ratio.alloc = 1
optimum.w = TRUE
b.prior = 1
seed = 134

opt.w = SPRToptimumW(n.sim = n.sim, n.event = n.event, n.start = n.start,
                     n.increment = 5, ve.null = ve.null, ve.alt = ve.alt,
                     type1 = type1, power = power, ratio.alloc = ratio.alloc,
                     w = seq(0.8, 0.9, by = 0.1), b.prior = b.prior, seed = seed)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

results = outputSPRT(method = method, n.sim = n.sim, n.event = n.event, n.start = n.start,
                    n.increment = n.increment, ve.null = ve.null, ve.alt = ve.alt,
                    type1 = type1, power = power, ratio.alloc = ratio.alloc,
                    optimum.w = optimum.w, w = opt.w$weight, b.prior = b.prior, seed = seed)
results
#> $boundary
#>   n_s r1 r2
#> 1  30  5 11
#> 2  35  7 13
#> 3  40  9 14
#> 4  45 10 16
#> 5  50 12 17
#> 6  55 13 19
#> 7  60 15 20
#> 
#> $cumulative.prob
#>   n_s r1 r2 P(RejectH0|H0) P(RejectH0|H1)
#> 1  30  5 11         0.0042         0.4246
#> 2  35  7 13         0.0084         0.6080
#> 3  40  9 14         0.0134         0.7438
#> 4  45 10 16         0.0141         0.7748
#> 5  50 12 17         0.0165         0.8425
#> 6  55 13 19         0.0166         0.8587
#> 7  60 15 20         0.0174         0.8928
#> 
#> $type1.power
#>          prob_dec
#> Under.H0   0.0174
#> Under.H1   0.8928
#> 
#> $prob.VE
#>   n_s P(x<=r1|H0) P(x>r2|H0) P(x<=r1|H1) P(x>r2|H1)
#> 1  30      0.0042     0.7576      0.4246     0.0273
#> 2  35      0.0080     0.7500      0.5976     0.0152
#> 3  40      0.0107     0.8377      0.7298     0.0217
#> 4  45      0.0058     0.8244      0.7148     0.0123
#> 5  50      0.0085     0.8867      0.8140     0.0144
#> 6  55      0.0045     0.8819      0.8019     0.0091
#> 7  60      0.0054     0.9223      0.8687     0.0116
#> 
#> $percent.stop
#>   n_s Avg(%)|H0 Avg(%)|H1
#> 1  30    0.7618    0.4519
#> 2  35    0.0461    0.1855
#> 3  40    0.0688    0.1428
#> 4  45    0.0217    0.0321
#> 5  50    0.0342    0.0699
#> 6  55    0.0114    0.0168
#> 7  60    0.0560    0.1010
#> 
#> $summary.stat
#>         Under.H0 Under.H1
#> Min.     30.0000  30.0000
#> 1st Qu.  30.0000  30.0000
#> Median   30.0000  35.0000
#> Mean     33.8930  37.6850
#> 3rd Qu.  30.0000  40.0000
#> Max.     60.0000  60.0000
#> sd        8.3049   9.8209
#> 
#> $time
#>          Time (in seconds)
#> Under.H0          8.895806
#> Under.H1          6.748705
#> 
#> $method
#> [1] "Bayes"
#> 
#> $n.event
#> [1] 60
#> 
#> $n.start
#> [1] 30
#> 
#> $n.increment
#> [1] 5
#> 
#> $ve.null
#> [1] 0.3
#> 
#> $ve.alt
#> [1] 0.75
#> 
#> $type1
#> [1] 0.025
#> 
#> $power
#> [1] 0.9
#> 
#> $ratio.alloc
#> [1] 1
#> 
#> $optimum.w
#> [1] TRUE
#> 
#> $w
#> [1] 0.8
#> 
#> $a.prior
#> [1] 0.5858209
#> 
#> $b.prior
#> [1] 1
#> 
#> $seed
#> [1] 134
```

<!-- What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so: -->
<!-- ```{r cars} -->
<!-- summary(cars) -->
<!-- ``` -->
<!-- You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. -->
<!-- You can also embed plots, for example: -->
<!-- ```{r pressure, echo = FALSE} -->
<!-- plot(pressure) -->
<!-- ``` -->
<!-- In that case, don't forget to commit and push the resulting figure files, so they display on GitHub! -->


NEWS.md
# Copyright (C) 2021 Merck Sharp & Dohme Corp. a subsidiary of Merck & Co., Inc., Kenilworth, NJ, USA.
#
# This file is part of bayesiansprt.
#
#     bayesiansprt is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     bayesiansprt is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with bayesiansprt.  If not, see <https://www.gnu.org/licenses/>.

# Workflow derived from https://github.com/r-lib/actions/tree/master/examples
# Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help

# bayesiansprt 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# Copyright (C) 2021 Merck Sharp & Dohme Corp. a subsidiary of Merck & Co., Inc., Kenilworth, NJ, USA.
#
# This file is part of bayesiansprt.
#
#     bayesiansprt is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     bayesiansprt is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with bayesiansprt.  If not, see <https://www.gnu.org/licenses/>.

# Workflow derived from https://github.com/r-lib/actions/tree/master/examples
# Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help

library(testthat)
library(bayesiansprt)
library(dplyr)

test_check("bayesiansprt")


tests/testthat/test-boundary.R
# Copyright (C) 2021 Merck Sharp & Dohme Corp. a subsidiary of Merck & Co., Inc., Kenilworth, NJ, USA.
#
# This file is part of bayesiansprt.
#
#     bayesiansprt is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     bayesiansprt is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with bayesiansprt.  If not, see <https://www.gnu.org/licenses/>.

# Workflow derived from https://github.com/r-lib/actions/tree/master/examples
# Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help

# setwd("/home/pauleri/bayesiansprt/tests")

test_that("Testing boundary in SPRTboundary", {

 a <- SPRTboundary(method = "Bayes", n.event = 60, n.start = 30, n.increment = 5,
                    ve.null = 0.3, ve.alt = 0.75, type1 = 0.025, power = 0.9,
                    ratio.alloc = 1, optimum.w = TRUE, w = 0.6, a.prior, b.prior = 1)

 b <- SPRTboundary(method = "Bayes", n.event = 60, n.start = 30, n.increment = 5,
                   ve.null = 0.3, ve.alt = 0.75, type1 = 0.025, power = 0.9,
                   ratio.alloc = 1, optimum.w = TRUE, w = 0.6, a.prior, b.prior = 1)

 expect_identical(a, b)
})


R/SPRTonVE.R
# Copyright (C) 2021 Merck Sharp & Dohme Corp. a subsidiary of Merck & Co., Inc., Kenilworth, NJ, USA.
#
# This file is part of bayesiansprt.
#
#     bayesiansprt is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     bayesiansprt is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with bayesiansprt.  If not, see <https://www.gnu.org/licenses/>.

# Workflow derived from https://github.com/r-lib/actions/tree/master/examples
# Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help

############################
# SPRT on Vaccine Efficacy #
############################

###########
# Library #
###########

library(ggplot2)
library(reshape2)
library(gridExtra)
library(lemon)
library(gtable)
library(grid)
library(lattice)
library(xtable)
library(dplyr)


# This function returns the following simulation results based on
# efficacy and futility boundaries:
# (1) boundary: specified event size and corresponding efficacy and futility boundaries
# (2) prob.rejH0: probability of rejecting H0 under H0 (if ve = ve.null) or H1 (if ve = ve.alt)
# (3) cum.prob: cumulative probabilities -
#               total, reject H0, accept H0 under H0 (if ve = ve.null) or H1 (if ve = ve.alt)
# (4) prob.each.n: probabilities that x <= efficacy and x >= futility boundaries
#                  under H0 (if ve = ve.null) or H1 (if ve = ve.alt)
# (5) percent.stop: Percentage of average sample size for each specified event
# (6) summay.stat: Summary statistics of events: min, Q1, Q2, mean, Q3, max, sd
# for Frequentist and Bayesian SPRT
# (7) time: time to run the code

SPRTonVE = function(n.sim = 100000, boundary.val, ve.null, ve.alt, ve, ratio.alloc,
                    seed){
  #n.sim: total number of simulation, by default 100,000
  #boundary.val: Boundaries corresponding to required events;
  #              This values can be generated from SPRT.BayesFreq.boundary.R;
  #              We can specify the interim timepoints and get the boundaries
  #              by extracting "Boundary" results, or we can get the optimized
  #              interim time points and get the corresponding boundaries by
  #              extracting "Boundary.interim" results
  #ve.null: vaccine efficacy under null
  #ve.alt: vaccine efficacy under alternative
  #ve: vaccine efficacy based on the model, if under H0, ve = ve.null; if under H1, ve = ve.alt
  #ratio.alloc: Treatment vs placebo ratdomization = K:1, provide K
  #seed.val: seed value to simulate data from binomial, by default 134

  start.time = Sys.time()
  #No of events
  n.count = nrow(boundary.val)

  #Probabilities of binomial success
  p = ratio.alloc*(1-ve)/(ratio.alloc*(1-ve)+1)

  ##################
  # Simulated data #
  ##################
  x =  matrix(NA, n.sim, n.count)

  for(i in 1:n.sim){
    set.seed(seed+3*i)
    x[i, 1] = rbinom(1, boundary.val$m[1], p)
    for(j in 2: n.count){
      set.seed(123*j+i)
      y =  rbinom(1, (boundary.val$m[j] - boundary.val$m[j-1]), p)
      x[i,j] = x[i, j-1] + y
    }
  }

  #####################################
  # Desicion based on SPRT boundaries #
  #####################################
  decision = matrix(NA, n.sim, n.count)
  for(i in 1:n.sim){
    for(j in 1: n.count){
      if(x[i,j]<= boundary.val[j,2]){
        decision[i,j] = 1#"Accept H1"
      }else if(x[i,j]>= boundary.val[j, 3]){
        decision[i,j] = 2#"Accept H0"
      }else{
        decision[i,j] = 3#"Continue"
      }
    }
  }

  ####################################################################################
  # Calculation for  probability of x<=r1 and x >=r2 for each interim under H0 or H1 #
  ####################################################################################
  prob_r12 = matrix(NA, n.count, 2)
  for(j in 1:n.count){
    for(k in 1:2){
      prob_r12[j, k] = length(which(decision[,j]==k))/n.sim
    }
  }

  #################################################################
  #Calculation for cumulative probability and average sample size #
  #################################################################S
  dec_n = rep(NA, n.sim)
  dec = rep(NA, n.sim)
  nval = rep(NA, n.sim)
  min_declist = rep(NA, n.sim)
  for(i in 1:n.sim){
    if(length(unique(decision[i,]))==1){
      if((max(decision[i,]) == 3)&&(min(decision[i,]) == 3)){
        dec_n[i] = 999
        dec[i] =  3
        nval[i] = max(boundary.val$m)+1
      }else{
        #position of decision
        dec_n[i] =  min(which(decision[i,] == min(decision[i,])))# 1
        #decision
        dec[i] = decision[i,dec_n[i]]
        #stopping n
        nval[i] = boundary.val$m[dec_n[i]]
      }
    }else if(length(unique(decision[i,]))==2){
      #if(min(decision[i,])<= 2 && max(decision[i,])==3){
      #position of decision
      dec_n[i] =  min(which(decision[i,] == min(decision[i,])))# 1
      #decision
      dec[i] = decision[i,dec_n[i]]
      #stopping n
      nval[i] = boundary.val$m[dec_n[i]]
      #}
    }else if(length(unique(decision[i,]))==3){
      dec_n[i] = which(decision[i,] <3)[1]
      #decision
      dec[i] = decision[i,dec_n[i]]
      #stopping n
      nval[i] = boundary.val$m[dec_n[i]]
    }
  }

  #Calculation total probability under H0 or H1
  dec_nprob = rep(NA, n.count)
  for(j in 1:n.count){
    dec_nprob[j] = length(which(nval == boundary.val$m[j]))/n.sim
  }

  #Calculation probability of rejecting H0 under H0 or H1
  index1 = which(dec==1)
  dec_nprob1 = rep(NA, n.count)
  for(j in 1:n.count){
    dec_nprob1[j] = length(which(nval[index1] == boundary.val$m[j]))/n.sim
  }

  #Calculation probability of accepting H0 under H0 or H1
  index2 = which(dec==2)
  dec_nprob2 = rep(NA, n.count)
  for(j in 1:n.count){
    dec_nprob2[j] = length(which(nval[index2] == boundary.val$m[j]))/n.sim
  }


  prob_h1 = length(which(dec==1))/n.sim
  prob_h0 = length(which(dec==2))/n.sim
  prob_c = length(which(dec==3))/n.sim

  nval_nmax = nval
  nval_nmax[which(nval_nmax == (max(boundary.val$m)+1))] = max(boundary.val$m)
  Nval_summary = cbind(summary(nval_nmax))
  Nval_sd = sd(nval_nmax)

  #Type I error or power
  prob_dec = prob_h1

  #Average sample size
  percent.val = table(nval_nmax)/sum(table(nval_nmax))

  #Summary of the events
  summary_all = Nval_summary
  sd = round(Nval_sd, 4)

  cum.prob = round(cbind(cumsum(dec_nprob), cumsum(dec_nprob1), cumsum(dec_nprob2)), 4)
  colnames(cum.prob) = c("Total_cp",  "cp_rejectH0", "cp_acceptH0")

  summary.stat = data.frame(rbind(summary_all, sd))
  colnames(summary.stat) = c("Summary")

  prob.each.n = round(cbind(prob_r12), 4)
  colnames(prob.each.n) = c("x_less_r1", "x_greater_r2")

  percent.stop = round(cbind(percent.val), 4)
  colnames(percent.stop) = c("percent.stop")

  time.total = Sys.time() - start.time

  mylist = list(boundary = boundary.val, prob.rejH0 = data.frame(prob_dec),
                cum.prob = cum.prob, prob.each.n = prob.each.n,
                percent.stop = percent.stop, summary.stat = summary.stat,
                time = time.total, seed = seed)

  return(mylist)
}



R/SPRTboundary.R
# Copyright (C) 2021 Merck Sharp & Dohme Corp. a subsidiary of Merck & Co., Inc., Kenilworth, NJ, USA.
#
# This file is part of bayesiansprt.
#
#     bayesiansprt is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     bayesiansprt is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with bayesiansprt.  If not, see <https://www.gnu.org/licenses/>.

# Workflow derived from https://github.com/r-lib/actions/tree/master/examples
# Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help

############################################
# SPRT Threshold: Bayesian and Frequentist #
############################################

##############
# Hypothesis #
##############

# H0: ve = ve.null  vs H1: ve != ve.null (= ve.alt)
# where ve.null and ve.alt are the null and alternative values based on vaccine efficacy(ve)
# Similarly, we can write in terms of relative risk (phi)
# H0: phi = phi0 = 1 - ve.null  vs H1: phi = phi1 = 1 - ve.alt

###########
# Library #
###########

library(ggplot2)
library(reshape2)
library(gridExtra)
library(lemon)
library(gtable)
library(grid)
library(lattice)
library(xtable)
library(dplyr)

####################################################################
# Function of finding thresholds for Bayesian and Frequentist SPRT #
####################################################################

# This function returns the design specifications and the efficacy and futility
# boundaries based on the optimum interim for frequentist and Bayesian SPRT

SPRTboundary = function(method, n.event, n.start, n.increment = 1,
                        ve.null = 0, ve.alt, type1 = 0.025, power = 0.9, ratio.alloc = 1,
                        optimum.w = TRUE, w = 0.5, a.prior, b.prior = 1){

  if (w<0|w>1){
    stop('Weight should be between 0 to 1')
  }

  if (!method %in% c('Bayes', 'Freq')){
    stop('Method must be one of Bayes/Freq')
  }

  #method: Provide "Freq" for frequentist approach; "Bayes" for Bayesian approch
  #n.event: Total number of events
  #n.start: The event to start the study
  #n.increment: Differences between two interim, by default 1
  #ve.null: Vaccine efficacy under null
  #ve.alt: Vaccine efficacy under alternative
  #type1: Specified Type I error, by default 0.025
  #power: Specified power, by default 0.9
  #ratio.alloc: Treatment vs control randomization = M:1, provide M
  #optimum.w: TRUE if use optimum weight; FALSE if used provide the shape parameter a.prior
  #w: Weight to calculate shape of Beta prior if method = "Bayes" and optimum.w = TRUE
  #a.prior: Shape parameter of Beta prior if method = "Bayes" and optimum.w = FALSE
  #b.prior: Scale parameter of Beta prior if method = "Bayes", by default 1

  #Null and alternative values based on probabilities of binomial success
  p0 = ratio.alloc*(1-ve.null)/(ratio.alloc*(1-ve.null)+1)
  p1 = ratio.alloc*(1-ve.alt)/(ratio.alloc*(1-ve.alt)+1)

  #Calculating shape for Beta prior
  if(optimum.w == TRUE){
    p = w*p0 + (1-w)*p1
    a.prior = p/(1-p)
  }else if(optimum.w == FALSE){
    a.prior = a.prior
  }

  #Threshold values based on SPRT
  a.val = log(((1-power)/(1-type1)), base = exp(1))
  b.val = log(((power)/type1), base = exp(1))

  #Vector of events
  n.val = seq(n.start, n.event, by = n.increment)
  n.count = length(n.val)

  dec = vector("list", n.count)
  r1 = matrix(NA, n.count, 1)
  r2 = matrix(NA, n.count, 1)
  logTest = vector("list", n.count)
  for(k in 1:n.count){

    #Calculation of the Decision for Bayesian SPRT
    test.val = rep(NA, (n.val[k]+1))
    decision = rep(NA, (n.val[k]+1))
    x =  rep(NA, (n.val[k]+1))

    #Initialization for Bayesian
    m_0 = rep(NA, (n.val[k]+1))
    m_1 = rep(NA, (n.val[k]+1))
    a_post = rep(NA, (n.val[k]+1))
    b_post = rep(NA, (n.val[k]+1))

    #Initialization for Frequentist
    ratio =  rep(NA, (n.val[k]+1))

    #Loop for each n
    for(i in 1:(n.val[k]+1)){
      x[i] = i-1

      if(method == "Bayes"){
        #Posterior parameters
        a_post[i] = a.prior + x[i]
        b_post[i] = b.prior + n.val[k] - x[i]

        #Marginal distributions
        m_0[i] = choose(n.val[k],  x[i])*beta(a_post[i], b_post[i])*(1-pbeta(p0, a_post[i], b_post[i]))/(beta(a.prior, b.prior)*(1-pbeta(p0, a.prior, b.prior)))
        m_1[i] = choose(n.val[k],  x[i])*beta(a_post[i], b_post[i])*(pbeta(p1, a_post[i], b_post[i]))/(beta(a.prior, b.prior)*(pbeta(p1, a.prior, b.prior)))

        #Test value based on bayes factor
        test.val[i] = log(m_1[i], base = exp(1)) - log(m_0[i], base = exp(1))
      }else if(method == "Freq"){
        #Likelihood ratio
        ratio[i] = (p1^x[i]*(1-p1)^(n.val[k]-x[i]))/(p0^x[i]*(1-p0)^(n.val[k]-x[i]))
        #Test value based on LR
        test.val[i] = log(ratio[i], base = exp(1))
      }else{
        print("Not available for this function")
      }

      #Decision based on threshold of SPRT
      if(test.val[i] > b.val){
        decision[i] = 1#"Accept H1"
      }else if(test.val[i] < a.val){
        decision[i] = 2#"Accept H0"
      }else{
        decision[i] = 3#"Continue"
      }
    }

    logTest[[k]] = cbind(n.val[k], x, test.val, decision)
    colnames(logTest[[k]]) = c("n", "x", "Test.value", "Decision")

    # Futility bound
    r2[k] = min(logTest[[k]][which(logTest[[k]][,4] == 2),2])

    # Efficacy bound
    r1cond = logTest[[k]][which(logTest[[k]][,4] == 1),2]
    Stop_Eff0 = pbinom(logTest[[k]][,2], n.val[k], p0)

    # Control interim alpha;
    index.cond = x[which(n.val[k] < .6*n.event & Stop_Eff0>(ve.alt-ve.null)*type1)]
    index.del = which(r1cond %in% intersect(index.cond, r1cond))
    r1[k] = ifelse(length(index.del == "TRUE")>0, max(r1cond[-index.del]), max(r1cond))
  }

  boundary = data.frame(n.val, r1, r2)
  colnames(boundary) = c("m", "lower", "upper")

  boundary.list = list(boundary = boundary, method = method, n.event = n.event,
                       n.start = n.start, n.increment = n.increment, ve.null = ve.null,
                       ve.alt = ve.alt, type1 = type1, power = power,
                       ratio.alloc = ratio.alloc, optimum.w = optimum.w, w = w,
                       a.prior = a.prior, b.prior = b.prior)

  return(boundary.list)
}



R/SPRTresult.R
# Copyright (C) 2021 Merck Sharp & Dohme Corp. a subsidiary of Merck & Co., Inc., Kenilworth, NJ, USA.
#
# This file is part of bayesiansprt.
#
#     bayesiansprt is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     bayesiansprt is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with bayesiansprt.  If not, see <https://www.gnu.org/licenses/>.

# Workflow derived from https://github.com/r-lib/actions/tree/master/examples
# Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help

###################################
# Combined results for SPRT on VE #
###################################

###########
# Library #
###########

library(ggplot2)
library(reshape2)
library(gridExtra)
library(lemon)
library(gtable)
library(grid)
library(lattice)
library(xtable)
library(dplyr)


# This function returns the following simulation results based on
# efficacy and futility boundaries:
# (1) boundary: No of events and corresponding efficacy and futility boundaries
# (2) type1.power: Probability of rejecting H0 under H0 (if ve = ve.null) or H1 (if ve = ve.alt)
# (3) cum.prob: Cumulative probabilities (CP) under null and alternative -
#               number of events, efficacy and futility boundaries,
#               total CP, CP for reject H0, CP for accept H0 under H0 (if ve = ve.null)
#               or H1 (if ve = ve.alt)
# (4) prob.all: Probabilities that x <= efficacy and x >= futility boundaries-
#               number of events, efficacy and futility boundaries
#               under H0 (if ve = ve.null) or H1 (if ve = ve.alt)
# (5) ass.percent: Percentage of average sample size for each specified event
#                  under null and alternative
# (6) summary.stat: Summary statistics of events: min, Q1, Q2, mean, Q3, max, sd
#                  under null and alternative
# (7) time.sprt: Time to run the code under null and alternative
# (8) weight.info: Optimum weight  and corresponding shape parameter of beta prior
#                  under null and alternative

outputSPRT = function(method, n.sim = 100000, n.event, n.start, n.increment, ve.null, ve.alt,
                      type1 = 0.025, power = 0.9, ratio.alloc = 1,  optimum.w = TRUE,
                      w = 0.5, a.prior, b.prior = 1, seed = 134){

  if (!method %in% c('Bayes', 'Freq')){
    stop('Method must be one of Bayes/Freq')
  }

  #method: Provide "Freq" for Frequentist approach; "Bayes" for Bayesian approach
  #n.sim: Simulation size, by default 100,000
  #n.event: Total number of events
  #n.start: The event to start the study
  #n.increment: Differences between two interim, by default 1
  #ve.null: Vaccine efficacy under null, by default 0
  #ve.alt: Vaccine efficacy under alternative
  #ratio.alloc: Treatment vs control randomization = M:1, provide M; by default 1
  #type1: Specified Type I error, by default 0.025
  #power: Specified power, by default 0.9
  #optimum.w: TRUE if use optimum weight; FALSE if used provide the shape parameter a.prior
  #w: Weight to calculate shape of Beta prior if method = "Bayes" and optimum.w = TRUE
  #a.prior: Shape parameter of Beta prior if method = "Bayes" and optimum.w = FALSE
  #seed: Seed value to simulate data from binomial, by default 134

  boundary.val = SPRTboundary(method = method, n.event = n.event, n.start = n.start,
                              n.increment = n.increment, ve.null = ve.null, ve.alt = ve.alt,
                              type1 = type1, power = power, ratio.alloc = ratio.alloc,
                              optimum.w =  optimum.w, w = w, a.prior = a.prior,
                              b.prior = b.prior)

  name.cases = c("Null", "Alternative")
  ve.val = c(ve.null, ve.alt)
  cases = data.frame(name.cases, ve.val)

  cum.prob.all = vector("list", nrow(cases))
  prob.all =  vector("list", nrow(cases))
  ass.percent.all =  vector("list", nrow(cases))
  summary.stat.all = matrix(NA, 7, nrow(cases))
  prob.rejH0.all = matrix(NA, nrow(cases), 1)
  weight.info = vector("list", nrow(cases))
  time.sprt =  vector("list", nrow(cases))

  for(l in 1:nrow(cases)){
    result = SPRTonVE(n.sim = n.sim, boundary.val = boundary.val$boundary, ve.null = ve.null,
                      ve.alt = ve.alt, ve = cases$ve.val[l], ratio.alloc = ratio.alloc,
                      seed = seed)

    val = result$percent.stop
    ass.val = data.frame(as.integer(rownames(val)), val)
    colnames(ass.val) = c("n_s", "ass")
    ass.val.complete =  data.frame(boundary.val$boundary$m)
    colnames(ass.val.complete) = c("n_s")
    library(dplyr)
    ass.all = left_join(ass.val.complete, ass.val, by = "n_s")
    ass.all[is.na(ass.all)] = 0
    ass.percent.all[[l]] = ass.all


    cum.prob.all[[l]] = data.frame(boundary.val$boundary, result$cum.prob)
    prob.all[[l]] = data.frame(boundary.val$boundary, result$prob.each.n)
    summary.stat.all[,l] = as.matrix(result$summary.stat)
    prob.rejH0.all[l,1] = as.matrix(unlist(result$prob.rejH0))

    time.sprt[[l]] = result$time

  }

  names(cum.prob.all) = cases$name.cases
  names(prob.all) = cases$name.cases
  names(ass.percent.all) = cases$name.cases
  colnames(summary.stat.all) = cases$name.cases
  rownames(summary.stat.all) = rownames(result$summary.stat)
  rownames(prob.rejH0.all) = cases$name.cases
  colnames(prob.rejH0.all) = rownames(as.matrix(unlist(result$prob.rejH0)))
  names(time.sprt) = cases$name.cases

  colnames(boundary.val$boundary) = c("n_s", "r1", "r2")

  indexcumprob0 = c(1:3, 5)
  indexcumprob1 = 5
  cumprob = data.frame(cum.prob.all$Null[,indexcumprob0], cum.prob.all$Alternative[,indexcumprob1])
  colnames(cumprob) = c("n_s", "r1", "r2", "P(RejectH0|H0)", "P(RejectH0|H1)")

  indexprob0 = c(1, 4:5)
  indexprob1 = 4:5
  prob.VE = data.frame(prob.all$Null[,indexprob0], prob.all$Alternative[,indexprob1])
  colnames(prob.VE) = c("n_s", "P(x<=r1|H0)", "P(x>r2|H0)", "P(x<=r1|H1)", "P(x>r2|H1)")

  colnames(summary.stat.all) = c("Under.H0", "Under.H1")

  ass.percent = data.frame(ass.percent.all$Null, ass.percent.all$Alternative[,2])
  colnames(ass.percent) = c("n_s", "Avg(%)|H0", "Avg(%)|H1")

  time = rbind(time.sprt$Null, time.sprt$Alternative)
  colnames(time) = "Time (in seconds)"
  rownames(time) = c("Under.H0", "Under.H1")

  rownames(prob.rejH0.all) =  c("Under.H0", "Under.H1")

  sim_outputF = list(boundary = boundary.val$boundary, cumulative.prob = cumprob,
                    type1.power = prob.rejH0.all, prob.VE = prob.VE,
                    percent.stop = ass.percent, summary.stat = summary.stat.all,
                    time = time, method = boundary.val$method,
                    n.event = boundary.val$n.event, n.start = boundary.val$n.start,
                    n.increment = boundary.val$n.increment, ve.null = boundary.val$ve.null,
                    ve.alt = boundary.val$ve.alt, type1 = boundary.val$type1,
                    power = boundary.val$power,
                    ratio.alloc = boundary.val$ratio.alloc,
                    seed = result$seed)

  sim_outputB = list(boundary = boundary.val$boundary, cumulative.prob = cumprob,
                     type1.power = prob.rejH0.all, prob.VE = prob.VE,
                     percent.stop = ass.percent, summary.stat = summary.stat.all,
                     time = time, method = boundary.val$method,
                     n.event = boundary.val$n.event, n.start = boundary.val$n.start,
                     n.increment = boundary.val$n.increment, ve.null = boundary.val$ve.null,
                     ve.alt = boundary.val$ve.alt, type1 = boundary.val$type1,
                     power = boundary.val$power, ratio.alloc = boundary.val$ratio.alloc,
                     optimum.w = boundary.val$optimum.w, w = boundary.val$w,
                     a.prior = boundary.val$a.prior, b.prior = boundary.val$b.prior,
                     seed = result$seed)

  if(method == "Bayes"){
    sim_output = sim_outputB
  }else if(method == "Freq"){
    sim_output = sim_outputF
  }

  return(sim_output)
}

